File: lib/ipc_server.py
#!/usr/bin/env python3
"""
LLM Vault - IPC Server
Handles communication between hooks and UI
Eventually this will communicate with the Swift menu bar app
For now, it provides a terminal-based interface
"""

import socket
import json
import os
import sys
import threading
from pathlib import Path
from credential_manager import CredentialManager


class VaultIPCServer:
    """IPC server for credential requests"""

    SOCKET_PATH = "/tmp/llm-vault.sock"

    def __init__(self):
        self.credential_manager = CredentialManager()
        self.socket = None

    def start(self):
        """Start the IPC server"""

        # Remove existing socket if present
        if os.path.exists(self.SOCKET_PATH):
            os.remove(self.SOCKET_PATH)

        # Create Unix domain socket
        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.socket.bind(self.SOCKET_PATH)
        self.socket.listen(5)

        # Make socket accessible
        os.chmod(self.SOCKET_PATH, 0o600)

        print(f"✓ LLM Vault IPC server started at {self.SOCKET_PATH}", flush=True)
        print("Waiting for credential requests from Claude Code hooks...\n", flush=True)

        try:
            while True:
                client, _ = self.socket.accept()
                # Handle each request in a separate thread
                thread = threading.Thread(target=self.handle_request, args=(client,))
                thread.daemon = True
                thread.start()

        except KeyboardInterrupt:
            print("\n✓ Server stopped", flush=True)
            self.stop()

    def handle_request(self, client_socket):
        """Handle a credential request from a hook"""
        try:
            # Receive request data
            data = client_socket.recv(4096)
            if not data:
                return

            request = json.loads(data.decode('utf-8'))
            action = request.get('action')

            print(f"\n📨 Received request: {action}", flush=True)

            if action == 'get_credential':
                response = self._handle_get_credential(request)
            elif action == 'store_credential':
                response = self._handle_store_credential(request)
            elif action == 'list_credentials':
                response = self._handle_list_credentials(request)
            elif action == 'delete_credential':
                response = self._handle_delete_credential(request)
            else:
                response = {'status': 'error', 'message': 'Unknown action'}

            # Send response
            client_socket.send(json.dumps(response).encode('utf-8'))

        except Exception as e:
            error_response = {'status': 'error', 'message': str(e)}
            client_socket.send(json.dumps(error_response).encode('utf-8'))

        finally:
            client_socket.close()

    def _handle_get_credential(self, request):
        """Handle get_credential request"""
        key_name = request.get('key')
        context = request.get('context', '')

        print(f"  🔑 Key: {key_name}", flush=True)
        print(f"  📝 Context: {context}", flush=True)

        # Try to get from keychain
        value = self.credential_manager.get_credential(key_name)

        if value:
            print(f"  ✓ Found in keychain", flush=True)
            return {
                'status': 'ok',
                'credential': value,
                'source': 'keychain'
            }

        # Not found - prompt user (for now, via terminal)
        print(f"\n🔐 Credential Required: {key_name}", flush=True)
        print(f"   Context: {context}", flush=True)
        print(f"   This will be shown in a popup in the menu bar app.", flush=True)

        # For MVP: Get from terminal input
        try:
            user_value = input(f"\n   Enter value for {key_name} (or press Enter to skip): ")

            if user_value:
                # Store in keychain
                metadata = {
                    'source': 'user_input',
                    'context': context,
                    'auto_inject': True
                }
                self.credential_manager.store_credential(key_name, user_value, metadata)

                print(f"   ✓ Stored in keychain", flush=True)

                return {
                    'status': 'ok',
                    'credential': user_value,
                    'source': 'user_input'
                }
            else:
                return {
                    'status': 'cancelled',
                    'message': 'User cancelled'
                }

        except (EOFError, KeyboardInterrupt):
            return {
                'status': 'cancelled',
                'message': 'User cancelled'
            }

    def _handle_store_credential(self, request):
        """Handle store_credential request"""
        key_name = request.get('key')
        value = request.get('value')
        metadata = request.get('metadata', {})

        success = self.credential_manager.store_credential(key_name, value, metadata)

        return {
            'status': 'ok' if success else 'error',
            'message': 'Stored' if success else 'Failed to store'
        }

    def _handle_list_credentials(self, request):
        """Handle list_credentials request"""
        creds = self.credential_manager.list_credentials()

        return {
            'status': 'ok',
            'credentials': creds
        }

    def _handle_delete_credential(self, request):
        """Handle delete_credential request"""
        key_name = request.get('key')

        success = self.credential_manager.delete_credential(key_name)

        return {
            'status': 'ok' if success else 'error',
            'message': 'Deleted' if success else 'Failed to delete'
        }

    def stop(self):
        """Stop the server"""
        if self.socket:
            self.socket.close()

        if os.path.exists(self.SOCKET_PATH):
            os.remove(self.SOCKET_PATH)


def request_credential(key_name: str, context: str = "", session_id: str = "") -> dict:
    """
    Client function to request a credential from the IPC server
    Called by hooks
    """
    try:
        # Connect to IPC server
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(VaultIPCServer.SOCKET_PATH)

        # Send request
        request = {
            'action': 'get_credential',
            'key': key_name,
            'context': context,
            'session_id': session_id
        }

        sock.send(json.dumps(request).encode('utf-8'))

        # Receive response
        response_data = sock.recv(4096)
        response = json.loads(response_data.decode('utf-8'))

        sock.close()

        return response

    except Exception as e:
        return {
            'status': 'error',
            'message': f'Failed to connect to LLM Vault server: {e}'
        }


if __name__ == '__main__':
    # Run the IPC server
    server = VaultIPCServer()
    server.start()

File: MenuBarApp/LocalLLMVault/LocalLLMVault/IPCServer.swift
//
//  IPCServer.swift
//  LLM Vault
//

import Foundation

protocol IPCServerDelegate: AnyObject {
    func credentialRequested(keyName: String, context: String, completion: @escaping (String?) -> Void)
}

class IPCServer {
    weak var delegate: IPCServerDelegate?
    private let socketPath = "/tmp/llm-vault.sock"
    private var serverSocket: Int32 = -1
    private var isRunning = false

    func start() {
        unlink(socketPath)

        serverSocket = socket(AF_UNIX, SOCK_STREAM, 0)

        guard serverSocket >= 0 else {
            print("✗ Failed to create socket")
            return
        }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)

        withUnsafeMutablePointer(to: &addr.sun_path.0) { ptr in
            socketPath.withCString { cString in
                strcpy(ptr, cString)
            }
        }

        let addrLen = socklen_t(MemoryLayout<sockaddr_un>.size)
        let bindResult = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                bind(serverSocket, $0, addrLen)
            }
        }

        guard bindResult >= 0 else {
            print("✗ Failed to bind socket")
            close(serverSocket)
            return
        }

        chmod(socketPath, 0o600)

        guard listen(serverSocket, 5) >= 0 else {
            print("✗ Failed to listen on socket")
            close(serverSocket)
            return
        }

        isRunning = true
        print("✓ IPC server listening on \(socketPath)")

        DispatchQueue.global(qos: .userInitiated).async {
            self.acceptConnections()
        }
    }

    func stop() {
        isRunning = false
        if serverSocket >= 0 {
            close(serverSocket)
            serverSocket = -1
        }
        unlink(socketPath)
    }

    private func acceptConnections() {
        while isRunning {
            let clientSocket = accept(serverSocket, nil, nil)

            guard clientSocket >= 0 else { continue }

            DispatchQueue.global(qos: .userInitiated).async {
                self.handleClient(socket: clientSocket)
            }
        }
    }

    private func handleClient(socket: Int32) {
        defer { close(socket) }

        var buffer = [UInt8](repeating: 0, count: 4096)
        let bytesRead = read(socket, &buffer, buffer.count)

        guard bytesRead > 0,
              let jsonData = Data(buffer.prefix(bytesRead)) as Data?,
              let request = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
              let action = request["action"] as? String else {
            return
        }

        print("📨 IPC request: \(action)")

        var response: [String: Any]

        switch action {
        case "get_credential":
            response = handleGetCredential(request)
        case "store_credential":
            response = handleStoreCredential(request)
        case "list_credentials":
            response = handleListCredentials()
        case "delete_credential":
            response = handleDeleteCredential(request)
        default:
            response = ["status": "error", "message": "Unknown action"]
        }

        if let responseData = try? JSONSerialization.data(withJSONObject: response),
           let responseBytes = [UInt8](responseData) as [UInt8]? {
            write(socket, responseBytes, responseBytes.count)
        }
    }

    private func handleGetCredential(_ request: [String: Any]) -> [String: Any] {
        guard let keyName = request["key"] as? String else {
            return ["status": "error", "message": "Missing key name"]
        }

        let context = request["context"] as? String ?? ""

        if let value = KeychainManager.shared.getCredential(keyName: keyName) {
            return [
                "status": "ok",
                "credential": value,
                "source": "keychain"
            ]
        }

        let semaphore = DispatchSemaphore(value: 0)
        var result: String?

        delegate?.credentialRequested(keyName: keyName, context: context) { value in
            result = value
            semaphore.signal()
        }

        let timeout = DispatchTime.now() + .seconds(120)
        let waitResult = semaphore.wait(timeout: timeout)

        if waitResult == .timedOut {
            return ["status": "error", "message": "Timeout waiting for user input"]
        }

        if let value = result {
            return [
                "status": "ok",
                "credential": value,
                "source": "user_input"
            ]
        } else {
            return ["status": "cancelled", "message": "User cancelled"]
        }
    }

    private func handleStoreCredential(_ request: [String: Any]) -> [String: Any] {
        guard let keyName = request["key"] as? String,
              let value = request["value"] as? String else {
            return ["status": "error", "message": "Missing key or value"]
        }

        let context = request["context"] as? String
        KeychainManager.shared.storeCredential(keyName: keyName, value: value, context: context)

        return ["status": "ok", "message": "Stored"]
    }

    private func handleListCredentials() -> [String: Any] {
        let credentials = KeychainManager.shared.listCredentials()

        let credentialDicts = credentials.map { credential in
            return [
                "key_name": credential.keyName,
                "context": credential.context ?? "",
                "created_at": ISO8601DateFormatter().string(from: credential.createdAt),
                "auto_inject": credential.autoInject
            ] as [String : Any]
        }

        return [
            "status": "ok",
            "credentials": credentialDicts
        ]
    }

    private func handleDeleteCredential(_ request: [String: Any]) -> [String: Any] {
        guard let keyName = request["key"] as? String else {
            return ["status": "error", "message": "Missing key name"]
        }

        KeychainManager.shared.deleteCredential(keyName: keyName)

        return ["status": "ok", "message": "Deleted"]
    }

    deinit {
        stop()
    }
}
